sslutils.c

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2002 - 2003
+ * NetGroup, Politecnico di Torino (Italy)
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Politecnico di Torino nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#ifdef HAVE_OPENSSL
+#include <stdlib.h>
+
+#include "portability.h"
+#include "sslutils.h"
+#include "pcap/pcap.h"
+
+int uses_ssl; //!< '1' to use TLS over the data socket
+char ssl_keyfile[PATH_MAX]; //!< file containing the private key in PEM format
+char ssl_certfile[PATH_MAX];  //!< file containing the server's certificate in PEM format
+char ssl_rootfile[PATH_MAX];  //!< file containing the list of CAs trusted by the client
+// TODO: a way to set ssl_rootfile from the command line, or an envvar?
+
+// TODO: lock?
+static SSL_CTX *ctx;
+
+static int ssl_init_once(int is_server, char *errbuf, size_t errbuflen)
+{
+  static int inited = 0;
+  if (inited) return 0;
+
+  SSL_library_init();
+  SSL_load_error_strings();
+  OpenSSL_add_ssl_algorithms();
+
+  SSL_METHOD const *meth = SSLv23_method();
```

Comment:
Data declarations after code needs a C99-compiler. Don't think that's a requirement in libpcap.

Reply:
We *do* require compiler support for at least some C99 features, including that one; we now require the same level of C99 support as Wireshark does:

> In general, not all C99 features can be used since some C compilers used to compile Wireshark, such as Microsoft's C compiler, don't support all C99 features.  The C99 features that can be used are:
> 
>  - flexible array members
>  - compound literals
>  - designated initializers
>  - "//" comments
>  - mixed declarations and code
>  - new block scopes for selection and iteration statements (that is, declaring the type in a for-loop like: for (int i = 0; i < n; i++) ;)
>  - macros with a variable number of arguments (variadic macros)
>  - trailing comma in enum declarations
>  - inline functions (guaranteed only by use of glib.h)

with "only by use of glib.h" replaced by "only by use of portability.h".

Our configure script doesn't enforce C99 support the way the Wireshark one does (in pre-3.0 releases; 3.0 requires CMake and doesn't have a configure script), by failing if it doesn't get a C99 compiler, but it should.

What I'm not sure of is why we're getting build errors with autotools *even though we're asking for a C99 compiler in the configure script*  and *even though it enables -std=gnu99* - perhaps the compiler is old enough that even gnu99 doesn't support mixed declarations and code.  I'll look into that.

Reply:
I've removed the -s flag from the make command, so it should print the commands being used at build time.

Reply:
We're getting those build errors *because we're asking for them* - we're building with `-Wdeclaration-after-statement`